# Assignment 1 

<h3> Objective </h3>
<p> To evaluate speedup derived from using multiple CPU cores via the multiprocessing facility in Python </p>
<h4> Step 1 </h4>
<p> Make a function which will utilize multiprocessing to execute another function </p>
<blockquote> This is essentially a wrapper<br> SYSTEM SPECS : Macbook Pro 1.4 GHz Quad-Core Intel Core i5 </blockquote>

In [1]:
# Generate a pool function which will accept a function as an argument and use a mapping with data recived to do computations
import time
import math
import multiprocessing
from multiprocessing import Pool

# A function for timing a job that uses a pool of processes.
#  f is a function that takes a single argument
#  data is an array of arguments on which f will be mapped
#  pool_size is the number of processes in the pool. 
def pool_process(f, data, pool_size):
    tp1 = time.time()
    pool = Pool(processes=pool_size) # initialize the Pool.
    result = pool.map(f, data);       # map f to the data using the Pool of processes to do the work 
#     print(result)
    pool.close() # No more processes
    pool.join()  # Wait for the pool processing to complete. 
#     print("Results", result)
    time_taken = int(time.time()-tp1)
    print("Overall Time:", time_taken)
    return time_taken


<h4> Step 2:</h4><br>
The function we will use for initial analysis to check perfomance hit.<blockquote> The function used here checks if the number passed to function is prime or not. The function will be called in a wrapper function which will pass a series of data and record time of execution. The process will be repeated with multiprocessing and analysis will be done</blockquote>

In [2]:
def check_prime(num):
    t1 = time.time()
    res = False
    if num > 0:
        # check for factors
        for i in range(2,num):
            if (num % i) == 0:
#                 print(num,"is not a prime number")
#                 print(i,"times",num//i,"is",num)
#                 print("Time:", int(time.time()-t1))
                break
        else:
#             print(num,"is a prime number")
#             print("Time:", time.time()-t1) 
            res = True
            # if input number is less than
            # or equal to 1, it is not prime
    return res

<h4> Step 3 : Get require data</h4><br>
<blockquote>
In order for analysis to be devoid of noise created by background process, context switching and scheduling we will need to run the program over a large set of number. To do this I took 3000 eight digit prime numbers from website provided and stored it in a txt file called 'primes.txt'. From this we will take subset of primes and will pass it to generate time required for exection </blockquote>

In [3]:
text_file = open("primes.txt", "r")

The txt file contains prime number in string format. Regex is used to extract the number and store it in a list

In [4]:
import re
regex = re.compile(r'[0-9]+')

In [5]:
lst = []
for oneline in text_file:
    lines = re.findall(regex,oneline)
    if(len(lines) > 0):
        for i in range(0,7):
            lst.append(int(lines[i]))                

Making data of diffrent size for analysis

In [6]:
dataRange_10 = lst[0:10]
dataRange_20 = lst[0:20]
dataRange_30 = lst[0:30]

In [7]:
dataRange_700 = lst[0:701]

<h4> Step 4 : Analysis </h4>
<blockquote> Here we start our analysis. For first part I will pass 700 large prime numbers to <b> pool_process </b> function and measure time taken to finish execution. This process will be repeated with multiple process from one process to 4 process and we will measure the time taken for each and later I will plot this in an excel sheet as an observation </blockquote> 

Single process - 700 primes

In [8]:
#pool_process(my_func, dataRange, 1)
time_taken_single_process_700 = pool_process(check_prime, dataRange_700, 1)

Overall Time: 591


This execution with just single process took 587 seconds to finish execution. We will do same test with multiple process below. I removed the verbose part of function as it holds no significance to the experiment and makes the notebook less readable

Two process - 700 primes

In [9]:
time_taken_dual_process_700 = pool_process(check_prime, dataRange_700, 2)

Overall Time: 306


The execution time almost halved when multi processing was enabled lets see results when we employ more processs

Three process - 700 primes

In [10]:
time_taken_triple_process_700 = pool_process(check_prime, dataRange_700, 3)

Overall Time: 228


It seems that the gains by increasing the process count is starting to plateau 

4 process - 700 primes

In [11]:
time_taken_quad_process_700 = pool_process(check_prime, dataRange_700, 4)

Overall Time: 195


In [12]:
time_taken_five_process_700 = pool_process(check_prime, dataRange_700, 5)

Overall Time: 191


Still improving but gains are modest. The results show that we might need to find an optimum number of process to handle a task. Rather than increasing the number of process as comprise between gain and process overhead.

In [13]:
time_taken_six_process_700 = pool_process(check_prime, dataRange_700, 6)

Overall Time: 200


We currently tested effect of multi-programming on 700 data points. Next we will see how it effects for much smalled number of datapoints and see if we can observe same trends. For that we will test with 20 and 30 prime numbers

<h6>Test with 20 primes</h6>

In [14]:
#pool_process(my_func, dataRange, 1)
time_taken_single_process_20 = pool_process(check_prime, dataRange_20, 1)

Overall Time: 17


In [15]:
#pool_process(my_func, dataRange, 2)
time_taken_double_process_20 = pool_process(check_prime, dataRange_20, 2)

Overall Time: 9


In [16]:
#pool_process(my_func, dataRange, 3)
time_taken_triple_process_20 = pool_process(check_prime, dataRange_20, 3)

Overall Time: 7


In [17]:
#pool_process(my_func, dataRange, 4)
time_taken_quad_process_20 = pool_process(check_prime, dataRange_20, 4)

Overall Time: 5


In [18]:
#pool_process(my_func, dataRange, 4)
time_taken_five_process_20 = pool_process(check_prime, dataRange_20, 5)

Overall Time: 5


In [19]:
#pool_process(my_func, dataRange, 4)
time_taken_six_process_20 = pool_process(check_prime, dataRange_20, 6)

Overall Time: 5


<blockquote> Observations<br>
    The plateauing of the gains are much more evident with smaller times. We indeed see an improvement in exectuion of process but the gains from it starts to decreases after certain number of process </blockquote>

<h6>Test with 10 primes</h6>

In [20]:
#pool_process(my_func, dataRange, 1)
time_taken_single_process_10 = pool_process(check_prime, dataRange_10, 1)

Overall Time: 9


In [21]:
#pool_process(my_func, dataRange, 2)
time_taken_double_process_10 = pool_process(check_prime, dataRange_10, 2)

Overall Time: 5


In [22]:
#pool_process(my_func, dataRange, 3)
time_taken_triple_process_10 = pool_process(check_prime, dataRange_10, 3)

Overall Time: 3


In [23]:
#pool_process(my_func, dataRange, 4)
time_taken_quad_process_10 = pool_process(check_prime, dataRange_10, 4)

Overall Time: 2


In [24]:
#pool_process(my_func, dataRange, 4)
time_taken_five_process_10 = pool_process(check_prime, dataRange_10, 5)

Overall Time: 2


In [25]:
#pool_process(my_func, dataRange, 4)
time_taken_six_process_10 = pool_process(check_prime, dataRange_10, 6)

Overall Time: 3


The three sets of data indeed give a similar trend. Providing sufficent data to back our claims of multiprocessing speeding up certain tasks.

<h3> Results of Observation <h3>

| INPUTS     	| Single Process 	| Two Process 	| Three Process 	| Four Process 	| Five Process 	| Six Process 	|
|------------	|----------------	|-------------	|---------------	|--------------	|--------------	|-------------	|
| 700 Primes 	|   591 seconds  	| 306 seconds 	|  228 seconds  	|  195 seconds 	|  191 seconds 	| 200 seconds 	|
| 20 Primes  	|   17 seconds   	|  9 seconds  	|   7 seconds   	|   5 seconds  	|   5 seconds  	|  5 seconds  	|
| 10 Primes  	|    9 seconds   	|  5 seconds  	|   3 seconds   	|   2 seconds  	|   2 seconds  	|  3 seconds  	|

<h1> Part 2 </h1>

<h4> In this section I will repeat the experiment with a function different from prime checker and see if we can reproduce the results </h4>

In Iterative approach we use a for loop to create the factorial series
AlGORITHM
<ul>
    <li>
Define function called factorial_iterative(x)
    </li>
    <li>
If we input 1 return directly 1
    </li>
    <li>
If number greater than 1 is sent as parameter make a for loop which will iterate from 1 to given number
Factorial calculated in variable called fact as fact = fact*number
    </li>
    <li>
Finally return the factorial calculated
    </li>
    </ul>

The reason for using this function is because its O(n) therfore we expect a linear trend and factorials can be an area where we can see gains in multiprocessing.

create data inputs 10000 inputs can be bit large but current pc can hadle it

In [27]:
data_range = [x for x in range (10000)]

In [28]:
def factorial_iterative(x):
    if (x==1):
        return 1
    else:
        fact = 1
    for z in range(1,x+1):
        fact=fact*z
    return fact

In [29]:
time_taken_single_process_10000_factorial = pool_process(factorial_iterative, data_range, 1)

Overall Time: 64


In [30]:
time_taken_double_process_10000_factorial = pool_process(factorial_iterative, data_range, 2)

Overall Time: 37


In [31]:
time_taken_triple_process_10000_factorial = pool_process(factorial_iterative, data_range, 3)

Overall Time: 26


In [32]:
time_taken_quad_process_10000_factorial = pool_process(factorial_iterative, data_range, 4)

Overall Time: 21


In [33]:
time_taken_five_process_10000_factorial = pool_process(factorial_iterative, data_range, 5)

Overall Time: 20


In [34]:
time_taken_six_process_10000_factorial = pool_process(factorial_iterative, data_range, 6)

Overall Time: 34


<h3> Results of Observation <h3>

| INPUTS            	| Single Process 	| Two Process 	| Three Process 	| Four Process 	| Five Process 	| Six Process 	|
|-------------------	|----------------	|-------------	|---------------	|--------------	|--------------	|-------------	|
| 10,000 Factorials 	|   64 seconds   	|  37 seconds 	|   26 seconds  	|  21 seconds  	|  20 seconds  	|  34 seconds 	|

<blockquote> <p> The observations remain same even for diffrent function. We see a large improvement in execution speed when we employ multiprocessing however as number of process increases there is a decrease in gain and reaches a point where increasing number of process doesn't increase perfomance and creates overheads. Therfore we need to see how many process we need to employ to properly get good gains. This is understandable as as process becomes more than number of available cores we need to do context switching </p></blockquote>

Better visualizations will be provided in accompanying documents